### PR TITLE
Implement REST service with CRUD operations for User

### PR DESCRIPTION
**Pull request by Vladlis**

### CODE REVIEW

File: webservices/rest/src/main/java/com/epam/rest/Application.java [hunk 1]
```diff
@@ -0,0 +1,16 @@
+package com.epam.rest;
+
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.EnableAutoConfiguration;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.context.annotation.ComponentScan;
+
+@SpringBootApplication
+@ComponentScan
+@EnableAutoConfiguration
```
Comment (alexander-lukyanchikov-epam) [inline webservices/rest/src/main/java/com/epam/rest/Application.java]: Actually `@SpringBootApplication` [includes both of rest annotations](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/SpringBootApplication.java#L51) :)

File: webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+import org.springframework.stereotype.Repository;
+
+import java.util.HashMap;
+import java.util.Map;
+
+@Repository
+public class UserDaoImpl implements UserDao {
+
+    private final Map<Integer, User> storage = new HashMap<>();
+    private int counter = 0;
```
Comment (alexander-lukyanchikov-epam) [inline webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: I understand that it is just mock repository implementation, but you know that it is not thread-safe, right? What would be a way to make it safe, with minimum performance impact?

File: webservices/rest/src/main/java/com/epam/rest/Application.java (hunk 1 above)
Comment (sqshq) [inline webservices/rest/src/main/java/com/epam/rest/Application.java]: Actually @SpringBootApplication [includes both of the rest annotations](https://github.com/spring-projects/spring-boot/blob/master/spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/SpringBootApplication.java#L51) :)

File: webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java (hunk 2 above)
Comment (sqshq) [inline webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: I understand that it is just mock repository implementation, but you know that it is not thread-safe, right? What would be a way to make it safe, with minimum performance impact?

**Review by sqshq — CHANGES_REQUESTED** [review]

File: webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java (hunk 2 above)
Comment (Vladlis) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: Replaced with AtomicInteger - guess it is the most effective way.

Comment (sqshq) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: That's right, and there is convenient method `incrementAndGet` for that case.

Comment (sqshq) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: What about HashMap? Is it thread-safe to use it here?

Comment (Vladlis) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: Nope :)
Replaced with SynchronizedMap.

Comment (sqshq) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: It is not very efficient way actually. `ConcurrentHashMap` uses sophisticated techniques to reduce the need for synchronization and allow parallel non-blocking read access by multiple threads. `SynchronizedMap` will block threads and degrade performance.

Comment (Vladlis) [reply webservices/rest/src/main/java/com/epam/rest/dao/UserDaoImpl.java]: Thanks, forgot about it.
As I see after a bit of googling it is a matter of a trade-off between consistency and performance. But as you asked about the minimum performance impact, then ConcurrentHashMap is preferable.
Replaced.

**Review by sqshq — APPROVED** [review]

**Pull request merged**